Gutenpack: Add/contact form to the gutenberg editor

enejb opened the pull request:
Fixes #9784

Simplifies the work done by @georgestephanis in https://github.com/Automattic/jetpack/pull/10256

By removing the js code since it should be automatically be included as part of the jetpack blocks. 
Now found here https://github.com/Automattic/wp-calypso/pull/27996 

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

### Todo: 
[ ] Rendering on the front-end (waiting on externals) - 
https://github.com/WordPress/gutenberg/pull/11334/  

#### Testing instructions:
* Go to https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=add/gutenblocks/contact-form&branch=add/contact-form-gutenblock-sdk

1. Connect Jetpack 
2. Visit the editor. Add the new contact form. 
3. Does it work as expected? 


#### Proposed changelog entry for your changes:
Add the contact form to the Gutenberg editor.

enejb committed:
Add the gutenberg contact form gutenberg blocks

enejb committed:
Gutenpack: Add contact form block

jeherve commented on modules/contact-form/grunion-contact-form.php:
```diff
@@ -232,6 +232,97 @@ function __construct() {
 		wp_register_style( 'grunion.css', GRUNION_PLUGIN_URL . 'css/grunion.css', array(), JETPACK__VERSION );
 		wp_style_add_data( 'grunion.css', 'rtl', 'replace' );
+
+		if ( function_exists( 'register_block_type' ) ) {
```
Should we use `Jetpack_Gutenberg::is_gutenberg_available` here?

enejb replied:
Yes good call. The plan is to convert things to jetpack_register_block calls which do that check as well.

jeherve reviewed (commented):
Not sure if this is ready for review yet, but I had a small question.

enejb committed:
Update form block to use contact-form

enejb committed:
Add do_blocks call to the contact render form

enejb committed:
Add register_contact_form_blocks block method

gititon committed:
Provide default labels

enejb committed:
Updates to fix tests

the default value of a lable should be null of the type is not set

enejb committed:
Minor fixes

enejb commented:
@georgestephanis Can you elaborate a bit more about the tests that you want to see?

georgestephanis commented:
Given post_content that it renders properly, primarily.

enejb committed:
Use the jetpack_register_block call to register the contact form blocks

Make sure we load them before we load the gutenberg blocks

enejb committed:
Bug fix: Only display selections when the option has a value.

enejb commented:
Also Fixes https://github.com/Automattic/wp-calypso/issues/28521

enejb committed:
Added tests that show what the form doesn't recieve a lavel that we defaul to the default label.

enejb committed:
Add tests that shows that we remove empty options when we pass an string containing the empty options

enejb committed:
Minor bug fix. Add closing div to the phone field.

enejb committed:
Add support for class added in gutenberg

enejb committed:
Refactored field Render method so that it is easier to follow.

More DRY field render method

enejb committed:
Minor fixes

enejb committed:
Add unitests for output.

enejb committed:
Unit Test fixes so that they work with older version of PHP

enejb committed:
Fix: Add more liberal url validaton

This removed the validation error when a user doesn't add a http:// or https:// infront of the url.

georgestephanis commented:
the failing tests seem to be coming from

```
Argument 1 passed to iterator_to_array() must implement interface Traversable, instance of DOMNodeList given, called in /tmp/wordpress-master/src/wp-content/plugins/jetpack/tests/php/modules/contact-form/test_class.grunion-contact-form.php on line 690 and defined
```

enejb committed:
Improved pattern for the url.

enejb committed:
Revert "Improved pattern for the url."

This reverts commit 105fccab7441dd3021adc28a5dd340416ca384e5.

enejb committed:
Revert "Fix: Add more liberal url validaton"

This reverts commit 0f1a6324042b3229b5418dff32184da268951933.

enejb committed:
Add support for the defaultValue block attribute

zinigor commented:
Sorry about spamming the PR, I'm testing a Fusion change that will fix status handling.

roccotripaldi committed:
Gutenberg: Conact form

ensure we are exposing the contact form's sub-blocks in block availability data.

the sub blocks will not be in blocks-manifest.json, so we'll need to ensure we
are also including blocks that have been registered internally within jetpack

enejb committed:
Fix the tests again :P

georgestephanis reviewed (commented):
Apart from some very minor stuff, very happy to clear this.

Thanks for the unit tests, refactoring how we're building the html into more manageable chunks, and the new assertions!  It looks super cool!

enejb committed:
Fixes indentation of the js code

enejb committed:
Remove comment

enejb committed:
Add php doc block for remove_empty

enejb committed:
Fixes ugly quotes.

enejb committed:
Fix the tests since we are expecting the double quotes now.

enejb committed:
With escaped quotes

jeherve commented:
Cherry-picked to `branch-6.8` in ea44f76c762b41dd8888eb9acf43dee34ab7c661

ockham commented:
It looks like the [Fusion-generated](https://github.com/Automattic/jetpack/pull/10537#issuecomment-435895770) counterpart patch (D20308-code) was never merged to WP.com, which is now causing trouble for a [PR](https://github.com/Automattic/jetpack/pull/10713) of mine that seeks to modify `class.jetpack-gutenberg.php`. Could someone who authored, reviewed, or merged this PR please take care of that? :grin:

ockham commented:
Oh, looks like D20308-code doesn't contain the changes that this PR made to `class.jetpack-gutenberg.php` (possibly because at the time it was merged, that file wasn't on Fusion's list yet). Maybe I can manually re-trigger patch creation...

ockham commented:
Fixed, thanks to @enejb's D21287-code.
